File: packages/preset/scripts/generate-shared.mjs
#!/usr/bin/env node
/**
 * generate-shared.mjs — Generate spec-driven shared templates + wrappers
 *
 * For each component in md3/components/:
 * 1. Read the original md3 source (most complete)
 * 2. Replace all $light-* → ${s('color.*')}, $dark-* → ${s('darkTokens.color.*')}
 * 3. Also replace $shape-* → ${s('shape.*')}
 * 4. Wrap in a make*Shortcuts(s) function
 * 5. Write to shared/components/Q*.unocss.ts
 * 6. Write thin wrappers to md3/components/, md2/components/, unstyled/components/
 */

import fs from 'fs'
import path from 'path'

const PRESET_DIR = new URL('..', import.meta.url).pathname
const MD3_DIR = path.join(PRESET_DIR, 'src/styles/md3/components')
const MD2_DIR = path.join(PRESET_DIR, 'src/styles/md2/components')
const SHARED_DIR = path.join(PRESET_DIR, 'src/styles/shared/components')
const UNSTYLED_DIR = path.join(PRESET_DIR, 'src/styles/unstyled/components')

// ─── Token mappings ────────────────────────────────────────────

const LIGHT_TOKENS = {
  primary: 'color.primary',
  'on-primary': 'color.onPrimary',
  'primary-container': 'color.primaryContainer',
  'on-primary-container': 'color.onPrimaryContainer',
  secondary: 'color.secondary',
  'on-secondary': 'color.onSecondary',
  'secondary-container': 'color.secondaryContainer',
  'on-secondary-container': 'color.onSecondaryContainer',
  tertiary: 'color.tertiary',
  'on-tertiary': 'color.onTertiary',
  'tertiary-container': 'color.tertiaryContainer',
  'on-tertiary-container': 'color.onTertiaryContainer',
  error: 'color.error',
  'on-error': 'color.onError',
  'error-container': 'color.errorContainer',
  'on-error-container': 'color.onErrorContainer',
  background: 'color.background',
  'on-background': 'color.onBackground',
  surface: 'color.surface',
  'on-surface': 'color.onSurface',
  'surface-variant': 'color.surfaceVariant',
  'on-surface-variant': 'color.onSurfaceVariant',
  'surface-container-lowest': 'color.surfaceContainerLowest',
  'surface-container-low': 'color.surfaceContainerLow',
  'surface-container': 'color.surfaceContainer',
  'surface-container-high': 'color.surfaceContainerHigh',
  'surface-container-highest': 'color.surfaceContainerHighest',
  outline: 'color.outline',
  'outline-variant': 'color.outlineVariant',
  'inverse-surface': 'color.inverseSurface',
  'inverse-on-surface': 'color.inverseOnSurface',
  'inverse-primary': 'color.inversePrimary',
  shadow: 'color.shadow',
  scrim: 'color.scrim',
  // Additional tokens (Quasar-specific)
  'dark-page': 'color.darkPage',
  brand: 'color.brand',
  accent: 'color.accent',
  positive: 'color.positive',
  negative: 'color.negative',
  info: 'color.info',
  warning: 'color.warning'
}

const SHAPE_TOKENS = {
  'corner-extra-small': 'shape.cornerExtraSmall',
  'corner-small': 'shape.cornerSmall',
  'corner-medium': 'shape.cornerMedium',
  'corner-large': 'shape.cornerLarge',
  'corner-extra-large': 'shape.cornerExtraLarge',
  'corner-full': 'shape.cornerFull'
}

const TYPOGRAPHY_TOKENS = {
  'font-family': 'typography.fontFamily',
  'display-large': 'typography.displayLarge',
  'display-medium': 'typography.displayMedium',
  'display-small': 'typography.displaySmall',
  'headline-large': 'typography.headlineLarge',
  'headline-medium': 'typography.headlineMedium',
  'headline-small': 'typography.headlineSmall',
  'title-large': 'typography.titleLarge',
  'title-medium': 'typography.titleMedium',
  'title-small': 'typography.titleSmall',
  'body-large': 'typography.bodyLarge',
  'body-medium': 'typography.bodyMedium',
  'body-small': 'typography.bodySmall',
  'label-large': 'typography.labelLarge',
  'label-medium': 'typography.labelMedium',
  'label-small': 'typography.labelSmall',
  'hover-opacity': 'typography.hoverOpacity',
  'focus-opacity': 'typography.focusOpacity',
  'pressed-opacity': 'typography.pressedOpacity',
  'dragged-opacity': 'typography.draggedOpacity'
}

/**
 * Convert $light-<kebab> and $dark-<kebab> to s() calls in template literals.
 * Handles both backtick strings and template expressions.
 */
function convertTokens(content) {
  let result = content

  // Build sorted patterns (longest first to avoid partial matches)
  const lightEntries = Object.entries(LIGHT_TOKENS).sort(
    (a, b) => b[0].length - a[0].length
  )
  const shapeEntries = Object.entries(SHAPE_TOKENS).sort(
    (a, b) => b[0].length - a[0].length
  )
  const typoEntries = Object.entries(TYPOGRAPHY_TOKENS).sort(
    (a, b) => b[0].length - a[0].length
  )

  // Pre-compile all replacements
  const replacements = []

  for (const [kebab, spec] of lightEntries) {
    const escaped = kebab.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
    replacements.push({
      pattern: new RegExp(`\\$light-${escaped}`, 'g'),
      replacement: `\${s('${spec}')}`
    })
    replacements.push({
      pattern: new RegExp(`\\$dark-${escaped}`, 'g'),
      replacement: `\${s('darkTokens.${spec}')}`
    })
  }

  for (const [kebab, spec] of shapeEntries) {
    const escaped = kebab.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
    replacements.push({
      pattern: new RegExp(`\\$shape-${escaped}`, 'g'),
      replacement: `\${s('${spec}')}`
    })
  }

  for (const [kebab, spec] of typoEntries) {
    const escaped = kebab.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
    replacements.push({
      pattern: new RegExp(`\\$typography-${escaped}`, 'g'),
      replacement: `\${s('${spec}')}`
    })
  }

  // Apply all replacements
  for (const { pattern, replacement } of replacements) {
    result = result.replace(pattern, replacement)
  }

  return result
}

/**
 * Convert a component name to a make function name.
 * E.g., "QBtn" → "makeQBtnShortcuts", "QAjaxBar" → "makeQAjaxBarShortcuts"
 */
function makeFnName(component) {
  return `make${component}Shortcuts`
}

/**
 * Convert the md3 component source into a shared template.
 */
function generateSharedTemplate(componentName, md3Source) {
  // Convert all $light-*/$dark-*/$shape-*/$typography-* references to s() calls
  let body = convertTokens(md3Source)

  // The source should already export `const shortcuts = [...]` or similar
  // We need to wrap it in a make*Shortcuts(s) function

  // Build the function name
  const fnName = makeFnName(componentName)

  // Check if the source has preflights
  const hasPreflights =
    body.includes('const preflights:') || body.includes('const preflights ')

  // Remove the original export statement
  body = body.replace(/^export \{ shortcuts \}\s*$/m, '')
  body = body.replace(/^export \{ preflights, shortcuts \}\s*$/m, '')
  body = body.replace(/^export \{ shortcuts, preflights \}\s*$/m, '')

  // Generate the result
  let result = `/**
 * ${componentName}.unocss.ts — Unified ${componentName} style template
 *
 * Spec-driven version: uses the bound spec resolver \`s()\` to
 * interpolate token values at shortcut-build time.
 *
 * Each style module calls \`${fnName}(bindSpec(spec))\` and
 * merges the result into its own shortcut list.
 *
 * @module Styles
 */

import type { Shortcut } from '@unocss/core'
import type { QuasarTheme } from '../../../theme.js'
import {
  componentCtxClass,
  componentClass,
  staticClass,
  qe
} from '../../_helpers.js'
import type { SpecResolver } from '../../_spec.js'

${body.trim()}

export function ${fnName}(s: SpecResolver): Shortcut<QuasarTheme>[] {
  return shortcuts
}
`

  if (hasPreflights) {
    result = `/**
 * ${componentName}.unocss.ts — Unified ${componentName} style template
 *
 * Spec-driven version: uses the bound spec resolver \`s()\` to
 * interpolate token values at shortcut-build time.
 *
 * Each style module calls \`${fnName}(bindSpec(spec))\` and
 * merges the result into its own shortcut list.
 *
 * @module Styles
 */

import type { Preflight, Shortcut } from '@unocss/core'
import type { QuasarTheme } from '../../../theme.js'
import {
  componentCtxClass,
  componentClass,
  staticClass,
  qe
} from '../../_helpers.js'
import type { SpecResolver } from '../../_spec.js'

${body.trim()}

export function ${fnName}(s: SpecResolver): Shortcut<QuasarTheme>[] {
  return shortcuts
}
`
  }

  return result
}

/**
 * Generate a thin wrapper file for a specific style.
 */
function generateWrapper(componentName, style, hasPreflights) {
  const fnName = makeFnName(componentName)
  const sharedPath = `../../shared/components/${componentName}.unocss.js`

  if (!hasPreflights) {
    return `import type { QuasarTheme } from '../../../theme.js'
import { ${fnName} } from '${sharedPath}'
import { bindSpec } from '../../_spec.js'

const s = bindSpec('${style}')
export const shortcuts = ${fnName}(s)
`
  }

  return `import { type Preflight, type Shortcut } from '@unocss/core'
import type { QuasarTheme } from '../../../theme.js'
import { ${fnName} } from '${sharedPath}'
import { bindSpec } from '../../_spec.js'

const s = bindSpec('${style}')
export const shortcuts = ${fnName}(s)
`
}

/**
 * Process all components.
 */
function processAll() {
  const files = fs.readdirSync(MD3_DIR).filter((f) => f.endsWith('.unocss.ts'))

  console.log(`Processing ${files.length} components...`)

  for (const file of files) {
    const match = file.match(/^Q(.+)\.unocss\.ts$/)
    if (!match) {
      console.log(`  SKIP (no Q match): ${file}`)
      continue
    }
    const componentName = `Q${match[1]}`
    const md3Path = path.join(MD3_DIR, file)
    const md2Path = path.join(MD2_DIR, file)
    const sharedPath = path.join(SHARED_DIR, file)
    const unstyledPath = path.join(UNSTYLED_DIR, file)

    // Read the md3 source (preferred source of truth)
    const md3Source = fs.readFileSync(md3Path, 'utf8')

    // Check if md2 exists
    const hasMd2 = fs.existsSync(md2Path)

    if (!hasMd2) {
      console.log(`  SKIP (no md2): ${file}`)
      continue
    }

    // Check for preflights
    const hasPreflights =
      md3Source.includes('const preflights:') ||
      md3Source.includes('const preflights ')

    // Generate shared template
    const sharedContent = generateSharedTemplate(componentName, md3Source)

    // Write shared template
    fs.mkdirSync(SHARED_DIR, { recursive: true })
    fs.writeFileSync(sharedPath, sharedContent)

    // Write md3 wrapper
    const md3Wrapper = generateWrapper(componentName, 'md3', hasPreflights)
    fs.writeFileSync(md3Path, md3Wrapper)

    // Write md2 wrapper
    if (hasMd2) {
      const md2Wrapper = generateWrapper(componentName, 'md2', hasPreflights)
      fs.writeFileSync(md2Path, md2Wrapper)
    }

    // Write unstyled wrapper if it exists
    if (fs.existsSync(unstyledPath)) {
      const unstyledWrapper = generateWrapper(
        componentName,
        'unstyled',
        hasPreflights
      )
      fs.writeFileSync(unstyledPath, unstyledWrapper)
    }

    console.log(`  OK: ${file}${hasPreflights ? ' (preflights)' : ''}`)
  }

  console.log('\nDone!')
}

processAll()

File: packages/preset/scripts/generate-shared.py
#!/usr/bin/env python3
"""
Generate shared component templates + style wrappers.

Reads md3 and md2 component files, creates shared templates in
src/styles/shared/components/ and replaces md3/md2 files with thin wrappers.

Strategy:
  Phase 1: Read ALL original files into memory first
  Phase 2: Generate shared templates from memory (never re-read)
  Phase 3: Write wrappers for md3 and md2  
  Phase 4: Verify by running tests
"""

import os
import re
import shutil
from pathlib import Path

BASE = '/home/stefan/Projects/unocss-preset-quasar/packages/preset/src/styles'
SHARED = f'{BASE}/shared/components'
MD3 = f'{BASE}/md3/components'
MD2 = f'{BASE}/md2/components'

os.makedirs(SHARED, exist_ok=True)


def read_file(path):
    with open(path) as f:
        return f.read()


def write_file(path, content):
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with open(path, 'w') as f:
        f.write(content)
    print(f"  Wrote {path}")


def camel_case(s):
    """Convert kebab-case to camelCase"""
    parts = s.split('-')
    return parts[0] + ''.join(p.capitalize() for p in parts[1:])


def component_name(filename):
    return filename.replace('.unocss.ts', '')


def make_func_name(comp):
    return f"make{comp}Shortcuts"


# ─── Categorization ──────────────────────────────────────────────

IDENTICAL = {
    'QAvatar', 'QBtnDropdown', 'QChatMessage', 'QExpansionItem',
    'QFile', 'QImg', 'QInnerLoading', 'QKnob', 'QPagination',
    'QParallax', 'QPopupEdit', 'QPullToRefresh', 'QRating',
    'QResponsive', 'QScrollarea', 'QSelect', 'QSeparator',
    'QSkeleton', 'QSpinner', 'QUploader', 'QVideo', 'QVirtualScroll'
}

DIFFERENT_VALUES = {
    'QAjaxBar', 'QBadge', 'QBanner', 'QBar', 'QCircularProgress',
    'QColorPicker', 'QFab', 'QFooter', 'QForm', 'QHeader', 'QIcon',
    'QInput', 'QIntersection', 'QOptionGroup', 'QPage', 'QPageSticky',
    'QSlider', 'QSpace', 'QSplitter', 'QTabPanel', 'QTimeline',
    'QToolbar', 'QTooltip', 'QTree'
}

DIFFERENT_STRUCTURE = {
    'QBtnToggle', 'QCard', 'QCarousel', 'QCheckbox', 'QChip',
    'QDate', 'QDialog', 'QDrawer', 'QEditor', 'QField', 'QItem',
    'QLayout', 'QLinearProgress', 'QMenu', 'QRadio', 'QSlideItem',
    'QStepper', 'QTable', 'QTabs', 'QTime', 'QToggle'
}

# Note: QBtn is ALREADY migrated - don't touch its files
# Note: QBreadcrumbs is different_structure but we handle it specially


def categorize(comp_name):
    if comp_name in IDENTICAL:
        return 'identical'
    elif comp_name in DIFFERENT_VALUES:
        return 'different_values'
    elif comp_name in DIFFERENT_STRUCTURE:
        return 'different_structure'
    return 'unknown'


# ─── Theme variable replacement ─────────────────────────────────

THEME_VAR_PATTERN = re.compile(
    r'\$(light|dark|shape-corner)-([a-zA-Z][a-zA-Z0-9-]*)'
)


def _theme_var_replacer(match):
    ns = match.group(1)
    name = match.group(2)
    camel = camel_case(name)
    
    if ns == 'light':
        path = f"color.{camel}"
    elif ns == 'dark':
        path = f"darkTokens.color.{camel}"
    elif ns == 'shape-corner':
        path = f"shape.corner{camel[0].upper() + camel[1:]}"
    else:
        return match.group(0)
    
    # Produce: ${s('path.here')}
    return f"${{{s_call(path)}}}"


def s_call(path):
    return f"s('{path}')"


def replace_theme_vars(text):
    """Replace $light-xxx, $dark-xxx, $shape-corner-xxx with s() calls."""
    return THEME_VAR_PATTERN.sub(_theme_var_replacer, text)


# ─── Shared template generators ──────────────────────────────────

def gen_shared_identical(comp_name, md3_content):
    """
    For identical components: wrap the original code in a make*Shortcuts function.
    The function body is the original file minus imports and export line.
    """
    func_name = make_func_name(comp_name)
    lines = md3_content.split('\n')
    
    # Extract imports, rewrite paths for shared location
    import_lines = []
    body_lines = []
    for line in lines:
        if line.startswith('import '):
            import_lines.append(line)
        elif 'export { shortcuts }' in line or 'export const shortcuts' in line:
            continue
        else:
            body_lines.append(line)
    
    body = '\n'.join(body_lines)
    
    # Build output using string concatenation to avoid f-string { } conflicts with JS
    parts = []
    parts.append(f"""/**
 * {comp_name}.unocss.ts — Shared {comp_name} style template
 *
 * Spec-driven version: uses the bound spec resolver `s()` to
 * interpolate token values at shortcut-build time.
 *
 * Each style module calls `{func_name}(bindSpec(spec))` and
 * merges the result into its own shortcut list.
 *
 * @module Styles
 */

""")
    for imp in import_lines:
        # Fix import paths - from '../../../theme.js' stays same for shared
        # from '../../_helpers.js' stays same for shared
        parts.append(imp + '\n')
    
    # Use a template without .format() to avoid KeyError from TypeScript { } in body
    template_lines = [
        "import type { SpecResolver } from '../../_spec.js'",
        '',
        '/**',
        f' * Generate {comp_name} shortcuts bound to a specific style\'s spec.',
        ' *',
        ' * @param s - The bound spec resolver from `bindSpec(spec)`.',
        ' * @returns An array of UnoCSS shortcuts.',
        ' */',
        f'export function {func_name}(s: SpecResolver): Shortcut<QuasarTheme>[] {{',
        indent_body(body, 2),
        '}',
        ''
    ]
    parts.append('\n'.join(template_lines))
    
    return ''.join(parts)


def gen_shared_different(comp_name, md3_content):
    """
    For components that differ: use the md3 version as base but replace
    UnoCSS theme variables with s() calls.
    """
    func_name = make_func_name(comp_name)
    lines = md3_content.split('\n')
    
    import_lines = []
    body_lines = []
    for line in lines:
        if line.startswith('import '):
            import_lines.append(line)
        elif 'export { shortcuts }' in line or 'export const shortcuts' in line:
            continue
        else:
            body_lines.append(line)
    
    # Join body then replace theme variables
    body = '\n'.join(body_lines)
    body = replace_theme_vars(body)
    
    parts = []
    parts.append(f"""/**
 * {comp_name}.unocss.ts — Shared {comp_name} style template
 *
 * Spec-driven version: uses the bound spec resolver `s()` to
 * interpolate token values at shortcut-build time.
 *
 * Each style module calls `{func_name}(bindSpec(spec))` and
 * merges the result into its own shortcut list.
 *
 * @module Styles
 */

""")
    for imp in import_lines:
        parts.append(imp + '\n')
    
    template_lines = [
        "import type { SpecResolver } from '../../_spec.js'",
        '',
        '/**',
        f' * Generate {comp_name} shortcuts bound to a specific style\'s spec.',
        ' *',
        ' * @param s - The bound spec resolver from `bindSpec(spec)`.',
        ' * @returns An array of UnoCSS shortcuts.',
        ' */',
        f'export function {func_name}(s: SpecResolver): Shortcut<QuasarTheme>[] {{',
        indent_body(body, 2),
        '}',
        ''
    ]
    parts.append('\n'.join(template_lines))
    
    return ''.join(parts)


def indent_body(body, spaces):
    """Indent every line of body by `spaces` spaces, except blank lines."""
    if not body.strip():
        return body
    indent = ' ' * spaces
    result = []
    for line in body.split('\n'):
        if line.strip():
            result.append(indent + line)
        else:
            result.append('')
    return '\n'.join(result)


# ─── Wrapper generators ─────────────────────────────────────────

def gen_wrapper(comp_name, style):
    """Generate a thin wrapper file."""
    func_name = make_func_name(comp_name)
    
    parts = []
    parts.append(f"""/**
 * {comp_name}.unocss.ts — {style} style wrapper
 *
 * Thin wrapper that binds the {style} spec to the shared template.
 *
 * @module Styles
 */

import {{ {func_name} }} from '../../shared/components/{comp_name}.unocss.js'
import {{ bindSpec, type SpecResolver }} from '../../_spec.js'

const s: SpecResolver = bindSpec('{style}')
export const shortcuts = {func_name}(s)
""")
    return ''.join(parts)


# ─── Main processing ─────────────────────────────────────────────

def main():
    # Phase 1: Read ALL files into memory first
    md3_files = sorted(os.listdir(MD3))
    
    print(f"Phase 1: Reading {len(md3_files)} md3 component files...")
    md3_originals = {}
    for filename in md3_files:
        if not filename.endswith('.unocss.ts'):
            continue
        md3_originals[filename] = read_file(f'{MD3}/{filename}')
    
    md2_originals = {}
    for filename in sorted(os.listdir(MD2)):
        if not filename.endswith('.unocss.ts'):
            continue
        md2_originals[filename] = read_file(f'{MD2}/{filename}')
    
    print(f"  Read {len(md3_originals)} md3 and {len(md2_originals)} md2 files")
    
    # Phase 2-3: Generate and write
    print(f"\nPhase 2-3: Generating shared templates and wrappers...")
    
    for filename in sorted(md3_originals.keys()):
        comp_name = component_name(filename)
        md3_content = md3_originals[filename]
        
        if filename not in md2_originals:
            print(f"  SKIP {filename}: no md2 version")
            continue
        
        md2_content = md2_originals[filename]
        is_identical = (md3_content == md2_content)
        cat = categorize(comp_name)
        
        print(f"\n  --- {comp_name} ({cat}) ---")
        
        # Skip QBtn - already migrated
        if comp_name == 'QBtn':
            print(f"    Already migrated (shared template exists)")
            continue
        
        # Generate shared template
        if is_identical or cat == 'identical':
            shared = gen_shared_identical(comp_name, md3_content)
        elif cat in ('different_values', 'different_structure'):
            shared = gen_shared_different(comp_name, md3_content)
        else:
            # Unknown: try different_values approach
            shared = gen_shared_different(comp_name, md3_content)
        
        # Write shared template
        shared_path = f'{SHARED}/{filename}'
        write_file(shared_path, shared)
        
        # Write md3 wrapper
        md3_wrapper = gen_wrapper(comp_name, 'md3')
        write_file(f'{MD3}/{filename}', md3_wrapper)
        
        # Write md2 wrapper
        md2_wrapper = gen_wrapper(comp_name, 'md2')
        write_file(f'{MD2}/{filename}', md2_wrapper)
        
        if cat == 'different_structure':
            print(f"    ⚠️  Structural differences - may need review")
    
    print(f"\n{'='*60}")
    print(f"Generation complete!")
    print(f"Shared templates: {SHARED}/")
    print(f"md3 wrappers: {MD3}/")
    print(f"md2 wrappers: {MD2}/")


if __name__ == '__main__':
    main()
